File: Design-Patterns-Using-Cpp/Visitor/VisitorShopping.cpp
#include <iostream>
#include <vector>

class Visitor; // Forward declaration

// Abstract base class for elements
class Element {
public:
    virtual double accept(Visitor& visitor) = 0;
    virtual double price() = 0;
    virtual ~Element() {}
};

// Concrete Book class
class Book : public Element {
private:
    double priceVar;
    int isbn;

public:
    Book(double price, int isbn) : priceVar(price), isbn(isbn) {}

    double price() override {
        return priceVar;
    }

    double accept(Visitor& visitor) override;
};

// Concrete Fruit class
class Fruit : public Element {
private:
    double priceVar;
    int quantity;
    std::string type;

public:
    Fruit(double price, int quantity, const std::string& type) : priceVar(price), quantity(quantity), type(type) {}

    double price() override {
        return priceVar;
    }

    double accept(Visitor& visitor) override;
};

// Abstract Visitor class
class Visitor {
public:
    virtual double visitBook(Book& book) = 0;
    virtual double visitFruit(Fruit& fruit) = 0;
    virtual ~Visitor() {}
};

// Concrete SundayDiscount class
class SundayDiscount : public Visitor {
public:
    double visitBook(Book& book) override {
        return book.price() - 50;
    }

    double visitFruit(Fruit& fruit) override {
        return fruit.price() - 5;
    }
};

// Concrete SaleDiscount class
class SaleDiscount : public Visitor {
public:
    double visitBook(Book& book) override {
        return book.price() / 2;
    }

    double visitFruit(Fruit& fruit) override {
        return fruit.price() / 2;
    }
};

// Implementation of accept for Book
double Book::accept(Visitor& visitor) {
    return visitor.visitBook(*this);
}

// Implementation of accept for Fruit
double Fruit::accept(Visitor& visitor) {
    return visitor.visitFruit(*this) * quantity;
}

// ShoppingCart class
class ShoppingCart {
private:
    std::vector<Element*> list;
    Visitor* visitor;

public:
    void add(Element* element) {
        list.push_back(element);
    }

    void setDiscountVisitor(Visitor* discount) {
        this->visitor = discount;
    }

    void accept() {
        double cost = 0;
        for (Element* element : list) {
            cost += element->accept(*visitor);
        }
        std::cout << "Total cost: " << cost << std::endl;
    }
};

// Client code
int main() {
    ShoppingCart cart;
    cart.add(new Fruit(100, 10, "Apple"));
    cart.add(new Book(100, 12345));

    cart.setDiscountVisitor(new SundayDiscount());
    cart.accept();

    cart.setDiscountVisitor(new SaleDiscount());
    cart.accept();

    return 0;
}

/*
Total cost: 1000
Total cost: 550
*/

File: Design-Patterns-Using-Swift/Visitor/VisitorShopping.swift
import Foundation

// Element protocol
protocol Element {
    func accept(visitor: Visitor) -> Double
    func price() -> Double
}

// Book class conforming to Element protocol
class Book: Element {
    private var _price: Double
    private var isbn: Int

    init(price: Double, isbn: Int) {
        self._price = price
        self.isbn = isbn
    }

    func price() -> Double {
        return _price
    }

    func accept(visitor: Visitor) -> Double {
        return visitor.visit(book: self)
    }
}

// Fruit class conforming to Element protocol
class Fruit: Element {
    private var _price: Double
    private var quantity: Int
    private var type: String

    init(price: Double, quantity: Int, type: String) {
        self._price = price
        self.quantity = quantity
        self.type = type
    }

    func price() -> Double {
        return _price
    }

    func accept(visitor: Visitor) -> Double {
        return visitor.visit(fruit: self) * Double(quantity)
    }
}

// Visitor protocol
protocol Visitor {
    func visit(book: Book) -> Double
    func visit(fruit: Fruit) -> Double
}

// SundayDiscount class conforming to Visitor protocol
class SundayDiscount: Visitor {
    func visit(book: Book) -> Double {
        return book.price() - 50
    }

    func visit(fruit: Fruit) -> Double {
        return fruit.price() - 5
    }
}

// SaleDiscount class conforming to Visitor protocol
class SaleDiscount: Visitor {
    func visit(book: Book) -> Double {
        return book.price() / 2
    }

    func visit(fruit: Fruit) -> Double {
        return fruit.price() / 2
    }
}

// ShoppingCart class
class ShoppingCart {
    private var list: [Element] = []
    private var visitor: Visitor?

    func add(element: Element) {
        list.append(element)
    }

    func setDiscountVisitor(discount: Visitor) {
        self.visitor = discount
    }

    func accept() {
        guard let visitor = visitor else {
            print("No discount visitor set.")
            return
        }

        var cost: Double = 0
        for element in list {
            cost += element.accept(visitor: visitor)
        }
        print("Total cost: \(cost)")
    }
}

// Client code.
let cart = ShoppingCart()
cart.add(element: Fruit(price: 100, quantity: 10, type: "Apple"))
cart.add(element: Book(price: 100, isbn: 12345))

cart.setDiscountVisitor(discount: SundayDiscount())
cart.accept()

cart.setDiscountVisitor(discount: SaleDiscount())
cart.accept()

/*
Total cost: 1000.0
Total cost: 550.0
*/
